File: shared/engine/logic.c
/**
 * @file logic.c
 * @brief Logic Engine - Logic Operations Implementation
 */

#include "logic.h"

/*============================================================================
 * Boolean Logic Functions
 *============================================================================*/

int32_t Logic_AND(const int32_t* inputs, uint8_t count)
{
    if (!inputs || count == 0) return 0;

    for (uint8_t i = 0; i < count; i++) {
        if (inputs[i] == 0) return 0;
    }
    return 1;
}

int32_t Logic_OR(const int32_t* inputs, uint8_t count)
{
    if (!inputs || count == 0) return 0;

    for (uint8_t i = 0; i < count; i++) {
        if (inputs[i] != 0) return 1;
    }
    return 0;
}

int32_t Logic_XOR(const int32_t* inputs, uint8_t count)
{
    if (!inputs || count == 0) return 0;

    uint8_t true_count = 0;
    for (uint8_t i = 0; i < count; i++) {
        if (inputs[i] != 0) true_count++;
    }
    return (true_count & 1);  /* Odd number of true inputs */
}

int32_t Logic_NAND(const int32_t* inputs, uint8_t count)
{
    return Logic_AND(inputs, count) ? 0 : 1;
}

int32_t Logic_NOR(const int32_t* inputs, uint8_t count)
{
    return Logic_OR(inputs, count) ? 0 : 1;
}

int32_t Logic_IsTrue(int32_t input)
{
    return (input != 0) ? 1 : 0;
}

int32_t Logic_IsFalse(int32_t input)
{
    return (input == 0) ? 1 : 0;
}

/*============================================================================
 * Comparison Functions
 *============================================================================*/

int32_t Logic_GT(int32_t a, int32_t b)
{
    return (a > b) ? 1 : 0;
}

int32_t Logic_GTE(int32_t a, int32_t b)
{
    return (a >= b) ? 1 : 0;
}

int32_t Logic_LT(int32_t a, int32_t b)
{
    return (a < b) ? 1 : 0;
}

int32_t Logic_LTE(int32_t a, int32_t b)
{
    return (a <= b) ? 1 : 0;
}

int32_t Logic_EQ(int32_t a, int32_t b)
{
    return (a == b) ? 1 : 0;
}

int32_t Logic_NEQ(int32_t a, int32_t b)
{
    return (a != b) ? 1 : 0;
}

/*============================================================================
 * Range Functions
 *============================================================================*/

int32_t Logic_InRange(int32_t value, int32_t min, int32_t max)
{
    return (value >= min && value <= max) ? 1 : 0;
}

int32_t Logic_OutsideRange(int32_t value, int32_t min, int32_t max)
{
    return (value < min || value > max) ? 1 : 0;
}

/*============================================================================
 * Generic Evaluate Function
 *============================================================================*/

int32_t Logic_Evaluate(LogicOp_t op, const int32_t* inputs, uint8_t count)
{
    if (!inputs || count == 0) return 0;

    switch (op) {
        /* Boolean operations */
        case LOGIC_OP_AND:
            return Logic_AND(inputs, count);

        case LOGIC_OP_OR:
            return Logic_OR(inputs, count);

        case LOGIC_OP_XOR:
            return Logic_XOR(inputs, count);

        case LOGIC_OP_NAND:
            return Logic_NAND(inputs, count);

        case LOGIC_OP_NOR:
            return Logic_NOR(inputs, count);

        case LOGIC_OP_IS_TRUE:
            return Logic_IsTrue(inputs[0]);

        case LOGIC_OP_IS_FALSE:
            return Logic_IsFalse(inputs[0]);

        /* Comparison operations (need 2 inputs) */
        case LOGIC_OP_GT:
            return (count >= 2) ? Logic_GT(inputs[0], inputs[1]) : 0;

        case LOGIC_OP_GTE:
            return (count >= 2) ? Logic_GTE(inputs[0], inputs[1]) : 0;

        case LOGIC_OP_LT:
            return (count >= 2) ? Logic_LT(inputs[0], inputs[1]) : 0;

        case LOGIC_OP_LTE:
            return (count >= 2) ? Logic_LTE(inputs[0], inputs[1]) : 0;

        case LOGIC_OP_EQ:
            return (count >= 2) ? Logic_EQ(inputs[0], inputs[1]) : 0;

        case LOGIC_OP_NEQ:
            return (count >= 2) ? Logic_NEQ(inputs[0], inputs[1]) : 0;

        /* Range operations (need 3 inputs: value, min, max) */
        case LOGIC_OP_RANGE:
            return (count >= 3) ? Logic_InRange(inputs[0], inputs[1], inputs[2]) : 0;

        case LOGIC_OP_OUTSIDE:
            return (count >= 3) ? Logic_OutsideRange(inputs[0], inputs[1], inputs[2]) : 0;

        default:
            return 0;
    }
}

File: shared/python/engine/logic.py
"""
Logic Engine - Boolean Logic Operations

Pure functions for boolean logic operations.
All functions return int32 values (0 or 1 for boolean, comparison results).
"""

from enum import IntEnum
from typing import Sequence


class LogicOp(IntEnum):
    """Logic operation types"""
    AND = 0
    OR = 1
    XOR = 2
    NAND = 3
    NOR = 4
    NOT = 5
    GT = 16
    GTE = 17
    LT = 18
    LTE = 19
    EQ = 20
    NEQ = 21
    IN_RANGE = 32
    OUTSIDE_RANGE = 33


def logic_and(inputs: Sequence[int]) -> int:
    """
    Logical AND of all inputs.
    Returns 1 if ALL inputs are non-zero.
    """
    if not inputs:
        return 0
    return 1 if all(x != 0 for x in inputs) else 0


def logic_or(inputs: Sequence[int]) -> int:
    """
    Logical OR of all inputs.
    Returns 1 if ANY input is non-zero.
    """
    if not inputs:
        return 0
    return 1 if any(x != 0 for x in inputs) else 0


def logic_xor(inputs: Sequence[int]) -> int:
    """
    Logical XOR of all inputs.
    Returns 1 if ODD number of inputs are non-zero.
    """
    if not inputs:
        return 0
    result = 0
    for x in inputs:
        if x != 0:
            result ^= 1
    return result


def logic_nand(inputs: Sequence[int]) -> int:
    """Logical NAND (NOT AND)"""
    return 1 if logic_and(inputs) == 0 else 0


def logic_nor(inputs: Sequence[int]) -> int:
    """Logical NOR (NOT OR)"""
    return 1 if logic_or(inputs) == 0 else 0


def logic_not(value: int) -> int:
    """Logical NOT"""
    return 0 if value != 0 else 1


def logic_gt(a: int, b: int) -> int:
    """Greater than comparison: a > b"""
    return 1 if a > b else 0


def logic_gte(a: int, b: int) -> int:
    """Greater than or equal: a >= b"""
    return 1 if a >= b else 0


def logic_lt(a: int, b: int) -> int:
    """Less than comparison: a < b"""
    return 1 if a < b else 0


def logic_lte(a: int, b: int) -> int:
    """Less than or equal: a <= b"""
    return 1 if a <= b else 0


def logic_eq(a: int, b: int) -> int:
    """Equality comparison: a == b"""
    return 1 if a == b else 0


def logic_neq(a: int, b: int) -> int:
    """Not equal comparison: a != b"""
    return 1 if a != b else 0


def logic_in_range(value: int, low: int, high: int) -> int:
    """Check if value is in range [low, high] inclusive"""
    return 1 if low <= value <= high else 0


def logic_outside_range(value: int, low: int, high: int) -> int:
    """Check if value is outside range [low, high]"""
    return 1 if value < low or value > high else 0


def logic_evaluate(op: LogicOp, inputs: Sequence[int]) -> int:
    """
    Evaluate logic operation.

    Args:
        op: Logic operation to perform
        inputs: Input values (meaning depends on operation)

    Returns:
        Operation result (0 or 1 for boolean ops)
    """
    if not inputs:
        return 0

    if op == LogicOp.AND:
        return logic_and(inputs)
    elif op == LogicOp.OR:
        return logic_or(inputs)
    elif op == LogicOp.XOR:
        return logic_xor(inputs)
    elif op == LogicOp.NAND:
        return logic_nand(inputs)
    elif op == LogicOp.NOR:
        return logic_nor(inputs)
    elif op == LogicOp.NOT:
        return logic_not(inputs[0])
    elif op == LogicOp.GT:
        return logic_gt(inputs[0], inputs[1]) if len(inputs) >= 2 else 0
    elif op == LogicOp.GTE:
        return logic_gte(inputs[0], inputs[1]) if len(inputs) >= 2 else 0
    elif op == LogicOp.LT:
        return logic_lt(inputs[0], inputs[1]) if len(inputs) >= 2 else 0
    elif op == LogicOp.LTE:
        return logic_lte(inputs[0], inputs[1]) if len(inputs) >= 2 else 0
    elif op == LogicOp.EQ:
        return logic_eq(inputs[0], inputs[1]) if len(inputs) >= 2 else 0
    elif op == LogicOp.NEQ:
        return logic_neq(inputs[0], inputs[1]) if len(inputs) >= 2 else 0
    elif op == LogicOp.IN_RANGE:
        return logic_in_range(inputs[0], inputs[1], inputs[2]) if len(inputs) >= 3 else 0
    elif op == LogicOp.OUTSIDE_RANGE:
        return logic_outside_range(inputs[0], inputs[1], inputs[2]) if len(inputs) >= 3 else 0

    return 0
